registry/image.go

```diff
@@ -0,0 +1,128 @@
+package registry
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+const (
+	dockerHubHost    = "index.docker.io"
+	dockerHubLibrary = "library"
+)
+
+// An image is a fully qualified name that refers to a particular image.
+// It is in the format: host[:port]/orgname/reponame[:tag]
+// Here, we refer to the "name" == orgname/reponame
+
+// Image can't really be a primitive string only, because we need to also
+// record information about it's creation time. (maybe more in the future)
+type Image interface {
+	WithTag(tag string) Image
+	WithCreatedAt(t *time.Time) Image
+	Components() (host, org, repo, tag string)
+	Host() string
+	Org() string
+	Repo() string
+	Tag() string
+	FQN() string
+	HostOrgRepo() string
+	OrgRepo() string
+	CreatedAt() *time.Time
+	Clone() Image
+}
+
+type image struct {
+	host, org, repo, tag string
+	createdAt            *time.Time `json:",omitempty"`
+}
+
+func ParseImage(s string, createdAt *time.Time) (Image, error) {
+	img := &image{
+		createdAt: createdAt,
+	}
+	parts := strings.Split(s, ":")
+	switch len(parts) {
+	case 0:
+		break
+	case 1:
+		break
+	case 2:
+		img.tag = parts[1]
+		s = parts[0]
+	default:
+		return nil, fmt.Errorf(`expected image name as either <image>:<tag> or just <image>`)
+	}
+	parts = strings.Split(s, "/")
+	switch len(parts) {
+	case 1:
+		img.host = dockerHubHost
+		img.org = dockerHubLibrary
+		img.repo = parts[0]
+	case 2:
+		img.host = dockerHubHost
+		img.org = parts[0]
+		img.repo = parts[1]
+	case 3:
+		img.host = parts[0]
+		img.org = parts[1]
+		img.repo = parts[2]
+	default:
+		return nil, fmt.Errorf(`expected image name as either "<host>/<org>/<image>", "<org>/<image>", or "<image>"`)
+	}
+	return img, nil
+}
+
+func (i *image) WithTag(t string) Image {
+	img := i.Clone().(*image)
+	img.tag = t
+	return img
+}
+
+func (i *image) WithCreatedAt(c *time.Time) Image {
+	img := i.Clone().(*image)
+	img.createdAt = c
+	return img
+}
```

Comment:
Ok, but this will break encapsulation rules. E.g. in the old version it was impossible to have an image without a name. Now it is possible because users have access to `img.Name`.